### PR TITLE
allow to set multiple meta blocks

### PR DESCRIPTION
**Pull request by doits**
Useful for includes and inheritance:

```rb
module SomeModule
  extend ActiveSupport::Concern

  included do
    meta do
      {
        module_meta: true
      }
    end
  end
end

class UserSerializer
  include FastJsonapi::ObjectSerializer
  include SomeModule

  meta do
    {
      user_meta: true
    }
  end
end

class EmployeeSerializer < UserSerializer
  meta do
    {
      employee_meta: true
    }
  end
end

EmployeeSerializer.new(employee).serializable_hash

{
  data: {
    [...]
    meta: {
      module_meta: true,
      user_meta: true,
      employee_meta: true
    }
  }
}
```

---

One thing to discuss is whether deep_merge or merge should be used. Only makes a difference in the exotic case of different meta blocks returning same keys though.

Otherwise I think this change makes composing serializers out of different modules and/or inheritance much more simpler, because each module or class can have meta blocks.

### CODE REVIEW

**Comment by stas** [comment]
@doits could you provide a bit more context here please? Based on your example, I'm having difficulties understanding why should we encourage such an approach instead of just using a plain base-class inheritance which is a pretty common design pattern (plus the benefits of using `super`). 

@christophersansone already provided a great example in https://github.com/fast-jsonapi/fast_jsonapi/issues/31#issuecomment-565551452, could you review it and let us know if there's a major reason why you'd not consider a design pattern explained above?

Thank you for the PR btw!

**Comment by christophersansone** [comment]
FWIW, from the standpoint of the DSL, I tend to think it's unclear how multiple `meta` blocks should behave.  Should it merge, or should it override?  Currently it overrides, and I would be hesitant to change that without offering some different syntax, or else it could break existing usage for those who are expecting it to work the original way.

One approach that would work and not even require a (breaking) syntax change would be this:

```ruby
  # input hash: {}
  # output hash: { module_meta: true }
  meta do |hash|
    hash.merge(module_meta: true)
  end

  # input hash: { module_meta: true }
  # output hash: { module_meta: true, user_meta: true }
  meta do |hash|
    hash.merge(user_meta: true)
  end

  # input hash would have been: { module_meta: true, user_meta: true }
  # output hash: { overridden: true }
  meta do
    # overrides all previous meta
    { overridden: true }
  end
```

Basically the `meta` block can offer a hash as an optional input parameter.  The block can choose to disregard it, merge, deep merge, check for an existing key, etc.  It provides more flexibility to the developer, the library does not need to decide how to handle duplicate keys (merge vs. deep merge vs. override), and the original syntax still works as is.

If @stas likes this approach, and if @doits can update the PR with this syntax with tests and an updated README, it's a 👍 .

**Comment by christophersansone** [comment]
Scratch that last comment.  The `meta` block is already offered `|record, params|` like all other blocks.  Maybe we could offer the `hash` as a third parameter?

I think this is something that can be done in userland as well, if it's not a common requirement (untested):

```ruby
class BaseSerializer
  include FastJsonapi::ObjectSerializer
  
  def self.multi_meta(&block)
    @multi_meta_blocks ||= []
    @multi_meta_blocks.push(block)
  end

  meta do
    if @multi_meta_blocks
      result = {}
      @multi_meta_blocks.each { |b| result.merge!(b.call) }
      return result
    end
  end
end

def UserSerializer < BaseSerializer
  multi_meta do
    { user_meta: true }
  end

  multi_meta do
    { employee_meta: true }
  end
end
```

**Comment by doits** [comment]
> could you provide a bit more context here please?

To give a (simplified) real world example for what I use it: Some serializers should output current user's permission in their meta in addition to serializer's own meta:

```rb
module Permissions
  extend ActiveSupport::Concern

  included do
    attribute :locked

    meta do |record|
      {
        permissions: {
          update: Current.user.can_update?(record),
          destroy: Current.user.can_destroy?(record)
        }
     }
    end
  end
end

class UserSerializer
  include FastJsonapi::ObjectSerializer
  include Permissions

  attribute :first_name

  meta do |record]
    {
      verified: record.verified?
    }
  end
end
```

This way included modules can simply add their own meta to other serializers without the need to change the serializers that include that module.

In a way it makes `meta` work like `attributes` works already: multiple calls *add* things but do not remove the ones defined before.

So mostly my use case is to be able to include modules like this and the inheritance change is simply to make it work consistently in both cases - which is a breaking change for both cases, that is right, I did not think about this yet.

Thinking about how `attribute` works there maybe `meta` could accept a param which is the top level key:

```rb
# adds or overwrites `meta[:permissions]`, other meta is unchanged
meta :permissions do
  {
     update: Current.user.can_update?(record),
     destroy: Current.user.can_destroy?(record)
  }
end

# overwrite whole `meta`, like existing behavior
meta do
  {
    some_meta: true
  }
end
```

So inheritance and including modules would work exactly like for `attribute` and it is backwards compatible, too.

>  Maybe we could offer the hash as a third parameter?

Yeah, this would work, too.

> I think this is something that can be done in userland as well

Thanks for that example, probably I will go with some monkey patching like this if this (or some similar way) will not be merged.

**Comment by doits** [comment]
I changed `meta` to work like described with a `param` like `attribute`, I now think this is cleaner and it keeps backwards compatibility, too. So unless you use `meta param do ... end`, everything should be the same as before.

I left the old commit for reference and can clean up / rebase later if you think this is something to merge, adding things to README, too.

**Comment by stas** [comment]
@doits I'm sorry but I'll have to close this. I don't see how this should be a concern of this library. We're trying to keep things simple and generic, and more importantly as strict as possible related to the JSONAPI spec challenges.

Your problem solution is easily solvable in the userland!
Based on your example, consider writing a method that does exactly what you need and passing it to the `meta` block.

So instead of having a separate call to build the Hash you need `meta` to return, consider this:
```ruby
module Permissions
  def self.current_user_can?(record)
    {
      verified: UserInteractor::Verified.call(user: record).verified?,
      permissions: {
        update: Current.user&.can_update?(record),
        destroy: Current.user&.can_destroy?(record)
      }
    }
  end
end

class UserSerializer
  include FastJsonapi::ObjectSerializer

  attribute :first_name

  meta do |record|
    Permissions.current_user_can?(record)
  end
end
```

Please close this issue if this suits your problem.

**Comment by doits** [comment]
OK, no problem, I can work around it myself (I do it already) and thought this might benefit the community.

The example you refactored misses one point though: Now the `Permissions` module can only be used for the `UserSerializer`, because `UserInteractor::Verified.call(user: record).verified?` makes only sense for an user. Of course the idea is to use the module in multiple interactors, I think I was not clear enough on this. So it must look like this:

```rb
module Permissions
  def self.current_user_can?(record)
    {
      permissions: {
        update: Current.user&.can_update?(record),
        destroy: Current.user&.can_destroy?(record)
      }
    }
  end
end

class UserSerializer
  include FastJsonapi::ObjectSerializer

  attribute :first_name

  meta do |record|
    {
      verified: UserInteractor::Verified.call(user: record).verified?
    }.merge(Permissions.current_user_can?(record))
  end
end

class OtherSerializer
  include FastJsonapi::ObjectSerializer

  meta do |record|
    Permissions.current_user_can?(record)
  end
end

class ThirdSerializer
  include FastJsonapi::ObjectSerializer

  meta do |record|
    { some: :meta }.merge(Permissions.current_user_can?(record))
  end
end
```

So basically what I wanted to avoid is to have to change every serializer's `meta` call when I have some generic logic by having modules I can include in multiple interactors. For attributes it is no problem already, but for meta it is not possible unless the respective meta call in the serializer is changed (or added if it didn't exist yet).

Of course it is possible to do it like this, only it feels cumbersome and not the ruby way IMO.

---

And the last point, besides the use case above and which came to me only yesterday, is the generic question: Why does `meta` work differently than `attribute`? In the end the define a hash that is outputted (`attributes: {}` vs `meta: {}`), so I would expect they have a similar call signature and behaviour. This PR right now is just a first step towards this (because then `meta` should at least respect `if: proc {}`, too), but maybe this makes more sense.

**Comment by doits** [comment]
For the record: I made `meta` work exactly like `attribute` in my fork (https://github.com/doits/fast_jsonapi/commit/1be4b07a02d22adc07cf6db45bf995c7d6e2acde), `meta` and `attribute` share the same logic and code how they are generated. Calling `meta` without a block is still possible to stay backwards compatible.

Feel free to think about the idea - if you came to the conclusion that this is worthwhile, I'd be happy to refactor the code (if there's something that doesn't fit) and go on with it until it can be merged. If not, no problem - maybe I will go on maintaining it just for me or do it differently in my projects then.

**Comment by christophersansone** [comment]
FWIW, I think this syntax has a lot of merit:

```ruby
meta :foo do |record, params|
  ...
end

meta :bar do |record, params|
  ...
end
```

Having a conditional `if` makes a lot of sense too, of course.  It definitely makes the `meta` syntax consistent with `attribute`.  And if a key is not specified, it can simply behave as it does today.  The logic would look something like this pseudocode:

```ruby
def self.meta(key = nil, &block)
  if key
    @meta_keys ||= {}
    @meta_keys[key] = block
  else
    @meta_without_key = block
  end
end

def self.serialize_meta(record, params)
  if @meta_without_key
    return @meta_without_key.call(record, params)
  elsif @meta_keys
    result = {}
    @meta_keys.each_pair { |key, block| result[key] = block.call(record, params) }
    return result
  end
end
```

This way, it will only execute the most recently defined `meta` that did not specify a key.  If all `meta` have keys specified, it will assemble the hash with each key defined.

Great job to @doits for iterating on the syntax!

I personally rarely use `meta`, so I have not run into these limitations, but this is an elegant solution worth consideration.  It would most likely be slower in most cases because of the iteration and multiple block calls as opposed to one block call that outputs a hash.  How much of a difference would have to be measured.  This gem needs to be true to its name (i.e. _fast_), so performance should be a factor on whether it should be supported out of the box and / or recommended as the right way to do it.

Lastly, I just also want to point out that this can be done in userland quite easily:

```ruby
class BaseSerializer
  include FastJsonapi::ObjectSerializer
  
  def self.meta_key(key, &block)
    @meta_keys ||= {}
    @meta_keys[key] = block
  end

  meta do |record, params|
    if @meta_keys
      result = {}
      @meta_keys.each_pair { |key, block| result[key] = b.call(record, params) }
      return result
    end
  end
end

def UserSerializer < BaseSerializer
  meta_key :user_meta do
    true
  end

  meta_key :employee_meta do
    true
  end
end
```

**Comment by doits** [comment]
FYI I already implemented it in my branch. I think reopening this PR should make it visible here then you can see the code more easily.

I can make some benchmarks, too, and see if I can squash out some more performance, but first we all must agree that this is a path worthwhile to follow.
